### PR TITLE
Feature/#23 menu game modes ranking

### LINKED ISSUE #23: UI Component | Join main menu, game modes, rules and score table
old hash: #19

**Acceptance Criteria**
- [ ] Component displays previously prepared components.
- [ ] The component displays the Star Wars Logo.
- [ ] The component displays the Main Menu.
- [ ] The component displays a sample image to be guessed.
- [ ] The component displays the rules of the game.
- [ ] The component displays two buttons.
- [ ] The right button is always "PLAY THE GAME".
- [ ] The left button is "Hall of fame", when pressed, it changes the view of the rules of the game to the ranking of players.
- [ ] After pressing the "Hall of fame" button, the Component displays the ranking of players in the selected game mode.
- [ ] After pressing the "Hall of fame" button it turns into the "Rules" button, pressing which will show the rules of the selected game mode again.
- [ ] Component performance covered by unit tests.

### PR DESCRIPTION
**Pull request by paula0403**
closes #23

### CODE REVIEW

File: src/app/components/MainMenu.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import createVisualImage from './VisualImage';
+import createLogo from './Logo';
+import createButtonRed from './ButtonRed';
+import createWhiteButtonWithIcon from './ButtonWhiteWithIcon';
+import createGameModeName from './GameModeName';
+import ModeMenu, { menuOption } from './ModeMenu';
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:6]: https://github.com/benmosher/eslint-plugin-import/blob/v2.22.1/docs/rules/no-cycle.md

Is it possible to solve it? 🔝

File: src/app/components/ModeMenu.js [hunk 2]
```diff
@@ -1,27 +1,36 @@
+import PageContent from './MainMenu';
```
Comment (Rednaxela700) [inline src/app/components/ModeMenu.js:1]: https://github.com/benmosher/eslint-plugin-import/blob/v2.22.1/docs/rules/no-cycle.md

Also here...

File: src/app/components/ModeRules.js [hunk 3]
```diff
@@ -1,9 +1,9 @@
...
-  people:
+  People:
     'You have one minute (1m) to answer as many questions as possible. During the game on each question you need to select who from Star Wars is showed on the left (Jar Jar Binks right now) from available options.',
-  vehicles:
+  Vehicles:
     'You have one minute (1m) to answer as many questions as possible. During the game on each question you need to select which vehicle from Star Wars is showed on the left.',
-  starships:
+  Starships:
     'You have one minute (1m) to answer as many questions as possible. During the game on each question you need to select which starship from Star Wars is showed on the left.',
 };
```
Comment (Rednaxela700) [inline src/app/components/ModeRules.js:9]: why?

Comment (Rednaxela700) [reply src/app/components/ModeRules.js:9]: if keys collide with eg variables names we can use destructuring like this
`const {people: newPeopleNameInCurrentScope} = rulesTexts;`

File: src/app/components/VisualImage.js [hunk 4]
```diff
@@ -2,25 +2,27 @@ function VisualImage(imageSource) {
...
 
     // Adding sourse to img
-    const img = document.querySelector('.visual-box__image');
-    img.setAttribute('src', imageSource);
-  } else {
-    throw new Error('Ups, format of input string is incorrect');
+    // const img = document.querySelector('.visual-box__image');
+    createImg.setAttribute('src', imageSource);
+    return createDiv;
   }
```
Comment (Rednaxela700) [inline src/app/components/VisualImage.js:22]: I think you can use the elementFactory function here

File: styles/components/_buttons.scss [hunk 5]
```diff
@@ -1,13 +1,15 @@
...
+  flex-grow: 1;
 
   // Display & Box Model
   border-radius: $border-radius;
   border: 0;
   padding: 1rem;
   box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
-  min-width: 400px;
+  max-width: 575px;
+  height: 100px;
```
Comment (Rednaxela700) [inline styles/components/_buttons.scss:12]: 👍

File: styles/components/_gameModeName.scss [hunk 6]
```diff
@@ -1,18 +1,17 @@
 .mode-info {
   background-color: $color-primary;
-  max-width: 70vw;
+  max-width: 1200px;
```
Comment (Rednaxela700) [inline styles/components/_gameModeName.scss:3]: 👍

File: src/app/components/ModeMenu.js [hunk 7]
```diff
@@ -1,27 +1,36 @@
...
+  // Creating nav's elements.
+  const newModeMenu = document.createElement('nav');
   const options = ['People', 'Vehicles', 'Starships'];
   options.forEach((option) => {
     const btn = document.createElement('button');
-    btn.classList.add('menu_button');
+    btn.classList.add('menu__button');
     btn.textContent = option;
-    menu.append(btn);
+    newModeMenu.appendChild(btn);
```
Comment (Rednaxela700) [inline src/app/components/ModeMenu.js:11]: elementFactory...

File: src/app/components/MainMenu.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+export const LoadPage = () => {
+  const quiz = document.querySelector('#swquiz-app');
+  const header = document.createElement('header');
+  const section = document.createElement('section');
+  const newNav = ModeMenu();
+  const gameLogo = createLogo();
+  header.classList.add('header');
+  section.classList.add('section');
+  quiz.append(header, section);
+  header.append(gameLogo, newNav);
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:74]: elementFactory

File: src/app/components/MainMenu.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  buttonsWrapper.classList.add('section__wrapper__buttons');
+  section.append(visualImage, gameWrapper);
+  gameWrapper.append(gameModeInfo, modeRules, buttonsWrapper);
+  buttonsWrapper.append(buttonRulesRanking, buttonPlay);
+
+  // Changing the rules and ranking view by pressing the button
+  let rules = false;
+  function handleChangeOfView() {
+    if (!rules) {
+      modeRules.remove();
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:39]: remember about the cleanup: 🧹 
https://stackoverflow.com/questions/36998877/what-is-the-difference-between-remove-and-removechild-method-in-javascript 👀

File: src/app/components/MainMenu.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      buttonRulesRanking.firstElementChild.classList.add(
+        'button__icon',
+        'fa',
+        'fa-id-badge',
+      );
+    }
+  }
+  buttonRulesRanking.addEventListener('click', handleChangeOfView);
+  return section;
+}
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:62]: try to avoid flags when possible. Maybe you could break it into smaller pieces? I see also some parts are being repeated, a function call might help to clean it up a bit

File: src/app/components/MainMenu.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    'fa-id-badge',
+  );
+  const gameModeInfo = createGameModeName(
+    `Who is this ${optionMode.selectOption}?`,
+  );
+  gameWrapper.classList.add('section__wrapper');
+  buttonsWrapper.classList.add('section__wrapper__buttons');
+  section.append(visualImage, gameWrapper);
+  gameWrapper.append(gameModeInfo, modeRules, buttonsWrapper);
+  buttonsWrapper.append(buttonRulesRanking, buttonPlay);
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:33]: call elementFactory()

File: src/app/components/MainMenu.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import createLogo from './Logo';
+import createButtonRed from './ButtonRed';
+import createWhiteButtonWithIcon from './ButtonWhiteWithIcon';
+import createGameModeName from './GameModeName';
+import ModeMenu, { menuOption } from './ModeMenu';
+import createModeRules from './ModeRules';
+
+// Creating page content
+function PageContent(optionMode = { selectOption: 'People' }) {
+  const section = document.querySelector('.section');
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:11]: Is it possible to grab this node without implicit DOM usage? It's not a mandatory change, consider it as a tip

File: src/app/components/MainMenu.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
+import createVisualImage from './VisualImage';
+import createLogo from './Logo';
+import createButtonRed from './ButtonRed';
+import createWhiteButtonWithIcon from './ButtonWhiteWithIcon';
+import createGameModeName from './GameModeName';
+import ModeMenu, { menuOption } from './ModeMenu';
+import createModeRules from './ModeRules';
+
+// Creating page content
+function PageContent(optionMode = { selectOption: 'People' }) {
```
Comment (Rednaxela700) [inline src/app/components/MainMenu.js:10]: is the fn argument supposed to contain any other data than `selectOption`?
If so, please use es6 destructuring, if not  simplify it setting a string instead

File: src/app/components/Logo.js [hunk 14]
```diff
@@ -1,15 +1,12 @@
...
-  elementA.setAttribute('href', './');
   elementDiv.classList.add('starwars__logo');
 
   // Adding to html structure
-  elementA.appendChild(elementDiv);
-  document.querySelector('#swquiz-app').appendChild(elementA);
+  return elementDiv;
 }
 
 export default Logo;
```
Comment (Rednaxela700) [inline src/app/components/Logo.js:12]: Is this file necessary ?

**Review by Rednaxela700 — CHANGES_REQUESTED** [review]
Didn't test it in working environment. Seems promising after some refactoring 🛠️

File: src/app/components/Logo.js (hunk 14 above)
Comment (DominikNowak) [reply src/app/components/Logo.js:12]: deleted

File: src/app/components/MainMenu.js (hunk 13 above)
Comment (DominikNowak) [reply src/app/components/MainMenu.js:10]: Changed for string

File: src/app/components/MainMenu.js (hunk 11 above)
Comment (DominikNowak) [reply src/app/components/MainMenu.js:33]: Refactored

File: src/app/components/MainMenu.js (hunk 8 above)
Comment (DominikNowak) [reply src/app/components/MainMenu.js:74]: Refactored

File: src/app/components/ModeMenu.js (hunk 7 above)
Comment (DominikNowak) [reply src/app/components/ModeMenu.js:11]: Refactored

File: src/app/components/ModeRules.js (hunk 3 above)
Comment (DominikNowak) [reply src/app/components/ModeRules.js:11]: Solved with using constants

File: src/app/components/VisualImage.js (hunk 4 above)
Comment (DominikNowak) [reply src/app/components/VisualImage.js:22]: @Enessetere did that in #88

File: src/app/components/MainMenu.js (hunk 9 above)
Comment (DominikNowak) [reply src/app/components/MainMenu.js:48]: Doesn't GC do that?

File: src/app/components/PageContent.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    buttonPlay,
+  );
+  const gameWrapper = elementFactory(
+    'div',
+    { className: 'section__wrapper' },
+    gameModeInfo,
+    modeRules,
+    buttonsWrapper,
+  );
+  section.ref.append(visualImage, gameWrapper);
```
Comment (vieraboschkova) [inline src/app/components/PageContent.js:80]: So you inject the elements again?

File: src/app/components/PageContent.js [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+  // Changing the rules and ranking view by pressing the button
+  function handleChangeOfView() {
+    if (section.buttonDisplay === RULES) {
+      section.buttonDisplay = RANK;
+      replaceContent(
```
Comment (vieraboschkova) [inline src/app/components/PageContent.js:86]: some parts seem repeatable, maybe could use something to make them default or another way?

**Review by vieraboschkova — COMMENTED** [review]
Let's analyze the need to render the elements again in the wrapper

File: src/app/components/ModeRules.js [hunk 17]
```diff
@@ -1,19 +1,23 @@
...
+    'which vehicle from Star Wars is showed on the left (Sith Speeder right now).',
+  [STARSHIPS]:
+    'which starship from Star Wars is showed on the left (TIE Advanced x1 right now).',
 };
 
 const ModeRules = (mode) => {
   const rulesTextElement = elementFactory(
     'div',
     { className: 'rules-text' },
-    rulesTexts[mode] ?? rulesTexts.people,
```
Comment (Rednaxela700) [inline src/app/components/ModeRules.js:16]: 😎

File: src/app/components/ModeRules.js [hunk 18]
```diff
@@ -1,19 +1,23 @@
...
+    `${repeatableText} ${rulesTexts[mode] ?? rulesTexts.people}`,
```
Comment (Rednaxela700) [inline src/app/components/ModeRules.js:20]: 😎

File: src/app/components/PageContent.js [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+import createButtonRed from './ButtonRed';
+import createWhiteButtonWithIcon from './ButtonWhiteWithIcon';
+import createGameModeName from './GameModeName';
+import createModeRules from './ModeRules';
+import createScoreTable from './ScoreTable';
+import { PEOPLE, STARSHIPS, VEHICLES } from '../constants';
+import elementFactory from '../utils/elementFactory';
+
+const section = {};
+const RULES = '<span></span> Rules';
```
Comment (Rednaxela700) [inline src/app/components/PageContent.js:11]: Empty span is no the best solution, but we don’t have the time for it right now.

File: src/app/components/PageContent.js (hunk 15 above)
Comment (Rednaxela700) [reply src/app/components/PageContent.js:86]: To avoid same multiplying the same nodes you could implement a check in render function for current node existence and remove it if necessary

**Review by Rednaxela700 — APPROVED** [review]

**Review by Enessetere — APPROVED** [review]
Looks good to me.
